### ISSUE TITLE
Enable release process

### ISSUE THREAD

**Issue by phillbaker** [opensearch-project/terraform-provider-opensearch#8]
In order to create releases, we'll need the following set as [repository secrets](https://github.com/opensearch-project/terraform-provider-opensearch/settings/secrets/actions) for the github actions (I do not have permission):
1. GPG_PRIVATE_KEY
2. PERSONAL_ACCESS_TOKEN

The values for these will also need to be generated. The personal access token is generated by Github, but needs to have permissions on this repo. A GPG key needs to be generated and the private key added to that environment variable.

As well, we'll need an account created on the Hashicorp terraform registry, see https://registry.terraform.io/publish/provider.

cc @bbarani

**Comment by bbarani** [comment]
@peterzhuamazon @prudhvigodithi Can you look in to this request?

**Comment by phillbaker** [comment]
@bbarani @peterzhuamazon @prudhvigodithi can you confirm who will be owning this on the opensearch side? Thanks!

**Comment by prudhvigodithi** [comment]
Hey @phillbaker thanks for checking back, let me take a look and I will post an update soon, thanks.

**Comment by prudhvigodithi** [comment]
Hey @phillbaker while we work on this release setup, can you please take of this [comment](https://github.com/opensearch-project/terraform-provider-opensearch/pull/7#issuecomment-1306197297) as well, to remove the `elasticsearch` reference and rename to `opensearch`?
Thank you
@bbarani @peterzhuamazon @dblock

**Comment by phillbaker** [comment]
@prudhvigodithi happy to take a look, can you open a separate issue to track that work?

**Comment by prudhvigodithi** [comment]
Hey @phillbaker thanks for reaching back, i'm on vacation for next few days, I will coordinate with @peterzhuamazon and @bbarani to get the release process move forward.
Thank you

**Comment by emopti-cmeyer** [comment]
@prudhvigodithi @peterzhuamazon @bbarani @phillbaker is this terraform provider still going to be released to the public? It appears to be stalled out waiting for the OpenSearch team to provide the necessary keys for pushing a public release.

Would be awesome if we could start using this.

**Comment by arpitjindal97** [comment]
Still waiting for the release of this

**Comment by reneeckstein** [comment]
yes, we too. Since AWS offered an upgrade to OpenSearch 2.3 it will affect more and more users I guess.

**Comment by prudhvigodithi** [comment]
Hey I'm just back from my vacation, starting 2023 I will focus on having a release process using GitHub Actions, I have the steps handy from the [terraform document](https://developer.hashicorp.com/terraform/tutorials/providers/provider-release-publish), I will initially test it on my fork repo before adding the github action to this main repo, also @phillbaker going through the document I see we need to have a [provider documentation](https://developer.hashicorp.com/terraform/tutorials/providers/provider-release-publish#document-your-provider) to move forward, when you get a chance can you please take a look at this?

Also regrading the query in my [previous comment ](https://github.com/opensearch-project/terraform-provider-opensearch/issues/8#issuecomment-1322560156) do we have to remove the elasticsearch reference and rename to opensearch before proceeding with release? @bbarani @peterzhuamazon @phillbaker @jackson-theisen WDYT? based on this we can take a call if rename has to be done before release.

@emopti-cmeyer @arpitjindal97 @reneeckstein 

Thank you

**Comment by phillbaker** [comment]
@prudhvigodithi there seems to be some misunderstanding of what's required here.

1. Github actions is already configured to build and release the provider (this was ported from the prior provider). All that's required are: (a) the keys I mentioned in the original description, (b) enabling on the terraform registry.
2. Documentation will be build automatically, I forgot to commit it, I can run the command and do that, but it is not blocking work on generating or adding the keys.
3. Please take a look at the import PR in https://github.com/opensearch-project/terraform-provider-opensearch/pull/7, all functional renaming has been done. Elasticsearch may be mentioned in comments where it is appropriate as a reference but it is not used functionally. Please open a separate issue to track this as I'm not clear exactly what you want to be changed. In my opinion, there is no further renaming blocking release.

Can you provide an estimated timeline when you will be able to generate these keys?

**Comment by prudhvigodithi** [comment]
Hey @phillbaker thanks, 

Ref to [ELASTICSEARCH](https://github.com/search?q=repo%3Aopensearch-project%2Fterraform-provider-opensearch%20elasticsearch&type=code), I can some ref in comments, print statements and unit tests.

Just an Update, finally we now have a [service account](https://registry.terraform.io/namespaces/opensearch-ci-bot) user linked with Hashicorp terraform registry,  
From the [publisher guide](https://developer.hashicorp.com/terraform/registry/providers/publishing) we just need the `GPG_PRIVATE_KEY` and `PASSPHRASE` created and export the public key to the above account, this should be done by end of this week @phillbaker 
@bbarani @peterzhuamazon

**Comment by bbarani** [comment]
As @prudhvigodithi mentioned, it would be great if we can clean up ElasticSearch reference in the code especially around print statements (as mentioned below) to avoid confusion.

![image](https://user-images.githubusercontent.com/70038446/210889765-9d87d0cc-14b6-4ce4-b04c-9a6e0e4a8512.png)

**Comment by arpitjindal97** [comment]
@phillbaker Can you please fix the issues, we are eagerly waiting for it to be released

**Comment by phillbaker** [comment]
@arpitjindal97 have the keys been added for the release process? I'd like to see that we've made progress there.

**Comment by prudhvigodithi** [comment]
Hey all can you please review this PR https://github.com/opensearch-project/terraform-provider-opensearch/pull/17 to rename to OpenSearch and Dashboard.
Thank you

**Comment by prudhvigodithi** [comment]
Just to add one more point this project repo is created using go client https://github.com/olivere/elastic which is advised not use in production, the project has to be migrated to use the official supported OpenSearch go client https://github.com/opensearch-project/opensearch-go
@phillbaker @jackson-theisen @bbarani 
![Screen Shot 2023-01-31 at 10 18 03 AM](https://user-images.githubusercontent.com/28733332/215848117-6f689dde-464c-4554-8e74-dc3b0f1ded63.png)

**Comment by phillbaker** [comment]
@prudhvigodithi you're misunderstanding the message. The message is saying not to pin to master, but rather to use go modules to pin to a specific version. This provider is already pinning to a specific version in go modules:

> the project has to be migrated to use the official supported OpenSearch go client https://github.com/opensearch-project/opensearch-go

Can you please open a separate issue for this and elaborate why this is a blocker to releasing a version? While I agree that it is good to do, the client is an internal implementation detail of the provider, not something that will be expose to customers.

**Comment by xsnrg** [comment]
Any updates on making this provider available in the registry?

**Comment by prudhvigodithi** [comment]
Hey All thanks for the patience on this release process and apologize for the delay, in order for the provider to be available in terraform registry, few app sec approvals are required according to the proposal made https://github.com/opensearch-project/terraform-provider-opensearch/issues/26. Once the approval is given the terraform-provider-opensearch should be good to go with the release. Meanwhile any contributions for the [proposal](https://github.com/opensearch-project/terraform-provider-opensearch/issues/26) made are welcome :) 
@bbarani @CEHENKLE @dblock

**Comment by prudhvigodithi** [comment]
Hey All, also adding security-reviews process link https://github.com/opensearch-project/.github/blob/main/RELEASING.md#security-reviews. Please take a look.

**Comment by prudhvigodithi** [comment]
Hey Everyone, thanks for the consistent patience, at this point we have an update to share, finally we received all the required approval for this provider to release. At this Point i will working to finish the pending items and make sure this provider is released.

As the 1st step I'm trying to improve the release workflow and make it more secure the way how it pulls the GPG keys (Open PR: https://github.com/opensearch-project/terraform-provider-opensearch/pull/28), as part of this I see multiple blockers which I posted in PR comment https://github.com/opensearch-project/terraform-provider-opensearch/pull/28#issuecomment-1451040485.

Once I have the blockers finalized we should proceed with the release.

Thank you

**Comment by phillbaker** [comment]
@prudhvigodithi the GPG key is not uploaded to Github secretes. Who has permissions to do that?

**Comment by prudhvigodithi** [comment]
Hey @phillbaker thanks for checing, we will upload the keys shortly, in progress to distinguish a terraform account either by a bot user ([service account](https://registry.terraform.io/namespaces/opensearch-ci-bot) or with github [OpenSearch project organization ](https://github.com/opensearch-project).
Thank you

**Comment by prudhvigodithi** [comment]
Hey @phillbaker the GPG keys are added, how about we do a test Github release with beta flag? and later do a proper version tag release that can be added to the Hashicorp registry?

**Comment by prudhvigodithi** [comment]
Hey @phillbaker just checking back again if you ok, can you test a Github release before we get a proper Hashicorp registry account?
Thank you

**Comment by phillbaker** [comment]
The gpg keys work, see the release here:
https://github.com/opensearch-project/terraform-provider-opensearch/releases/tag/v1.0.0-beta.2

**Comment by prudhvigodithi** [comment]
Hey All we have the provider finally published https://registry.terraform.io/providers/opensearch-project/opensearch/latest 👍🎉
Thank you

**Comment by prudhvigodithi** [comment]
Hey @phillbaker can we now do a normal release? If you are ok can you please cut a tag?
Thank you 
@bbarani @dblock @peterzhuamazon

**Comment by xocasdashdash** [comment]
Hi everyone!
I'm trying to make use of the released opensearch provider and it's hitting a "bug" when connecting to an AWS opensearch. The version detection is not working correctly and it says:
```
 opensearch version 2.3.0 is older than 6.0.0 and is not supported, flavor: 0.
```

looking into the code i think it's failing to detect the correct version in provider.go. 

I'm raising the comment here as i see the project is getting close to have a proper release in the terraform registry. I've tried looking for usage examples to see if i'm doing something wrong but i'm failing to find any. 

Can you confirm if i'm missing any steps in order to get this working? this is my provider configuration:
```
provider "opensearch" {
  url                  = "https://${aws_elasticsearch_domain.opensearch.domain_endpoint_options[0].custom_endpoint}"
  aws_region           = data.aws_region.current.name
  aws_assume_role_arn  = aws_iam_role.root.arn
  version_ping_timeout = 10
  healthcheck          = false
}
```

**Comment by phillbaker** [comment]
@xocasdashdash this provider does not yet support opensearch 2.x. 

@prudhvigodithi @bbarani we're still missing documentation, I'd prefer to wait until we get that in to release a 1.x. I'll need a couple of days to finish that up - some of the previous documentation wasn't auto generated, so I'll need to move that over to the auto generation.

**Comment by libfitz** [comment]
@phillbaker you mentioned [here](https://github.com/phillbaker/terraform-provider-elasticsearch/issues/298#issuecomment-1407848445) that _this_ provider is the way forward, does it still not work correctly with 2.x?

**Comment by xocasdashdash** [comment]
@libfitz i managed to make it work with v2.3 of aws opensearch by setting the version to "7.8" (created roles, role_mappings and a snapshot repository without problems), would be nice to have an issue understanding what's the breaking changes between v1 and v2 and what's missing here in order to support v2

**Comment by prudhvigodithi** [comment]
Hey @libfitz @xocasdashdash can you please add all the details required to support 2.x version of OpenSearch on the [issue created](https://github.com/phillbaker/terraform-provider-elasticsearch/issues/298#issuecomment-1407848445), this will help other contributors aware the changes required and can also help the repo owners to prioritize the 2.x support.
Thank you

**Comment by prudhvigodithi** [comment]
> @prudhvigodithi @bbarani we're still missing documentation, I'd prefer to wait until we get that in to release a 1.x. I'll need a couple of days to finish that up - some of the previous documentation wasn't auto generated, so I'll need to move that over to the auto generation.

Thanks @phillbaker can you take care of this, also please let me know if required any help.
Thank you

**Comment by phillbaker** [comment]
Got sidetracked by a few things, should have a PR later today

**Comment by phillbaker** [comment]
See https://github.com/opensearch-project/terraform-provider-opensearch/pull/31

**Comment by bbarani** [comment]
Btw, feel free to join OpenSearch public channel for discussions as well - https://opensearch.org/slack.html
